### ISSUE TITLE
Warnings are not filtered in csr_matrix.sum()

### ISSUE THREAD

**Issue by dizcza** [scipy/scipy#12449]
Dear scipy community,

suffering for one hour on the application side, I realized that there is a noteworthy bug in scipy.

```python
import scipy.sparse as sps
import warnings

counts = [1]
row_ids = [0]
column_ids = [1]


csr_matrix = sps.csr_matrix((counts, (row_ids, column_ids)),
                            shape=(1, 2),
                            dtype=int)


for i in range(5):
    s = csr_matrix.sum()
    warnings.warn("This message should be printed only once", UserWarning)
```

Warnings:
```
/home/ulianych/PycharmProjects/elephant/ignored/scipy_warn_bug.py:17: UserWarning: This message should be printed only once
  warnings.warn("This message should be printed only once", UserWarning)
/home/ulianych/PycharmProjects/elephant/ignored/scipy_warn_bug.py:17: UserWarning: This message should be printed only once
  warnings.warn("This message should be printed only once", UserWarning)
/home/ulianych/PycharmProjects/elephant/ignored/scipy_warn_bug.py:17: UserWarning: This message should be printed only once
  warnings.warn("This message should be printed only once", UserWarning)
/home/ulianych/PycharmProjects/elephant/ignored/scipy_warn_bug.py:17: UserWarning: This message should be printed only once
  warnings.warn("This message should be printed only once", UserWarning)
/home/ulianych/PycharmProjects/elephant/ignored/scipy_warn_bug.py:17: UserWarning: This message should be printed only once
  warnings.warn("This message should be printed only once", UserWarning)
```

As the message says, the warning should be printed only once. Citing the official python documentation,

> Repetitions of a particular warning for the same source location are typically suppressed.

Adding `warnings.simplefilter("module", UserWarning)` or `once` doesn't help either.

Note that if you comment out the line where the sum of a matrix is computed, the warning is printed only once as it should.

#### Scipy/Numpy/Python version information

Master branch (`pip install -e .`):
```python
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.4.1 1.18.1 sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)
```
I also tried conda v1.5.0 - same behavior.

**Comment by WarrenWeckesser** [comment]
@dizcza, thanks for reporting the issue.  I just checked, and the development version (`scipy.__version__` is `'1.6.0.dev0+5f69fc7'`) also shows the problem.

**Comment by WarrenWeckesser** [comment]
This appears to be a Python issue, not an issue specific to SciPy.  The SciPy code uses the context manager [`warnings.catch_warnings`](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings) internally, and apparently this affects the global state of the `warnings` code in an unexpected way.  This example is pure Python, and it generates the `UserWarning` three times:

```
import warnings

for i in range(3):
    with warnings.catch_warnings(record=True):
        pass
    warnings.warn("This message should be printed only once", UserWarning)
```

`catch_warnings` is [documented to modify global state](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings), but the effect seen here seems to be a bug.  Because `catch_warnings` modifies global state, it is not thread-safe (as explained in the documentation), but the examples here are single threaded.

**Comment by WarrenWeckesser** [comment]
There is a Python bug report for this issue: https://bugs.python.org/issue29672

**Comment by dizcza** [comment]
Good to know, thank you!
It has been opened for >3 years and does not seem to have any recent updates.
Please consider not using this buggy warnings catching, because otherwise on the application side we need either to ignore all the warnings or be annoyed with flooded messages that the user erroneously thinks that he has control over them by filtering the warnings with `once` (he doesn't).
Expect more issues of this sort.

I don't have any further comments, thank you for the quick feedback! You can close the issue as won't fix or leave it open if you think you may come up with a solution that avoids catching the warnings in the future.

**Comment by WarrenWeckesser** [comment]
@dizcza,  I agree that, given the Python bug, we should avoid using `catch_warnings`.  I'm not planning on closing the issue, but I don't know if other devs have already looked into this and found that this is the best we can do at the moment.
